src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1737,91 +1800,108 @@ public void ReturnHttp2Connection(Http2Connection connection, bool isNewConnecti
 
-                    if (!_http2RequestQueue.TryDequeueNextRequest(connection))
-                    {
-                        connection.ReleaseStream();
-                        break;
+                        // If the pool has been disposed of, we will dispose the connection below outside the lock.
+                        // We do this after processing the queue above so that any queued requests will be handled by existing connections if possible.
                     }
 
-                    if (NetEventSource.Log.IsEnabled()) connection.Trace("Dequeued waiting HTTP/2 request.");
-                }
-
-                // Since we only inject one connection at a time, we may want to inject another now.
-                CheckForHttp2ConnectionInjection();
-
-                if (_disposed)
-                {
-                    // If the pool has been disposed of, we want to dispose the connection being returned, as the pool is being deactivated.
-                    // We do this after the above in order to satisfy any requests that were queued before the pool was disposed of.
-                    Debug.Assert(_associatedHttp2ConnectionCount > (_availableHttp2Connections?.Count ?? 0));
-                    _associatedHttp2ConnectionCount--;
-                    poolDisposed = true;
-                }
-                else if (usable)
-                {
-                    if (_availableHttp2Connections is null)
+                    if (waiter is not null)
                     {
-                        _availableHttp2Connections = new List<Http2Connection>();
+                        Debug.Assert(!added);
+                        if (waiter.TrySetResult(connection))
+                        {
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Dequeued waiting HTTP2 request.");
+                            break;
+                        }
+                        else
+                        {
+                            Debug.Assert(waiter.Task.IsCanceled);
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Discarding canceled HTTP2 request from queue.");
+                            // Loop and process the queue again
+                        }
+                    }
+                    else
+                    {
+                        connection.ReleaseStream();
+                        if (added)
+                        {
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Put HTTP2 connection in pool.");
+                            return;
+                        }
+                        else
+                        {
+                            Debug.Assert(_disposed);
+                            if (NetEventSource.Log.IsEnabled()) connection.Trace("Disposing HTTP2 connection returned to pool. Pool was disposed.");
+                            connection.Dispose();
+                            return;
```

Comment:
We no longer correct the `_associatedHttp2ConnectionCount--` in this case - I take it that's fine since we're disposing everything anyway?

Reply:
Actually I think this is just an oversight. As you say it probably doesn't matter in practice, but we should still decrement the count here just in case. Thanks for catching this!